ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -1063,6 +1063,47 @@ long getLogUsedCapacity() {
     return getAbsolutePositionInLogForOffset(getCurrentEndOffset(indexSegments), indexSegments);
   }
 
+  /**
+   * @return absolute end position (in bytes) of latest PUT record when this method is invoked.
+   * @throws StoreException
+   */
+  long getAbsoluteEndPositionOfLastPut() throws StoreException {
+    IndexValue indexValueOfLastPut = null;
+    // 1. go through keys in journal in reverse order to see if there is a PUT index entry associated with the key
+    List<JournalEntry> journalEntries = journal.getAllEntries();
+    ConcurrentSkipListMap<Offset, IndexSegment> indexSegments = validIndexSegments;
+    // check entry in reverse order (from the most recent one) to find first PUT record in index
+    for (int i = journalEntries.size() - 1; i >= 0; --i) {
+      JournalEntry entry = journalEntries.get(i);
+      indexValueOfLastPut = findKey(entry.getKey(), new FileSpan(entry.getOffset(), getCurrentEndOffset(indexSegments)),
+          EnumSet.of(IndexEntryType.PUT), indexSegments);
+      if (indexValueOfLastPut != null) {
+        break;
+      }
+    }
+    if (!journalEntries.isEmpty() && indexValueOfLastPut == null) {
+      // 2. if not find in the journal, check index segments starting from most recent one (until latest PUT is found)
+      JournalEntry firstJournalEntry = journalEntries.get(0);
+      // generate a segment map in reverse order
+      ConcurrentNavigableMap<Offset, IndexSegment> segmentsMapToSearch =
+          indexSegments.subMap(indexSegments.firstKey(), true, indexSegments.lowerKey(firstJournalEntry.getOffset()),
+              true).descendingMap();
```

Comment:
nit: toInclusive can be false.

Reply:
It could be false in some cases but if journal doesn't cover the entire index segment (which means some entries slip out of journal that we didn't check). Explicitly set true will help avoid such edge case and ensure all entries in index segment are checked, although some entries would be checked twice.